contrib/http_util/src/main/java/io/opencensus/contrib/http/HttpServerHandler.java

```diff
@@ -0,0 +1,101 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opencensus.contrib.http;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import io.opencensus.trace.Annotation;
+import io.opencensus.trace.Span;
+import io.opencensus.trace.SpanBuilder;
+import io.opencensus.trace.SpanContext;
+import io.opencensus.trace.Tracer;
+import io.opencensus.trace.propagation.TextFormat;
+
+/**
+ * This helper class provides routine methods to instrument HTTP servers.
+ *
+ * @param <Q> the HTTP request entity.
+ * @param <P> the HTTP response entity.
+ * @since 0.13
+ */
+public class HttpServerHandler<Q, P> extends HttpHandler<Q, P> {
+
+  /**
+   * Creates a {@link HttpServerHandler} with given parameters.
+   *
+   * @param tracer the Open Census tracing component.
+   * @param extractor the {@code HttpExtractor} used to extract information from the
+   *     request/response.
+   * @param customizer the {@link HttpSpanCustomizer} used to customize span behaviors.
+   * @since 0.13
+   */
+  public HttpServerHandler(
+      Tracer tracer, HttpExtractor<Q, P> extractor, HttpSpanCustomizer<Q, P> customizer) {
+    super(tracer, extractor, customizer);
+  }
+
+  /**
+   * Instrument an incoming request before it is handled. Users should optionally invoke {@link
+   * #handleMessageReceived} after the request is received.
+   *
+   * <p>This method will create a span under the deserialized propagated parent context. If the
+   * parent context is not present, the span will be created under the current context.
+   *
+   * <p>This method will invoke {@link HttpSpanCustomizer#customizeSpanStart} after the span is
+   * successfully created.
+   *
+   * <p>The generated span will NOT be set as current context. User can use the returned value to
+   * control when to enter the scope of this span.
+   *
+   * @param <C> the type of the carrier
+   * @param textFormat the {@code TextFormat} used in HTTP propagation.
+   * @param getter the getter used when extracting information from the {@code carrier}.
+   * @param carrier the entity that holds the HTTP information.
+   * @param request the request entity.
+   * @return a span that represents the response handling process.
+   * @since 0.13
+   */
+  public <C> Span handleStart(
+      TextFormat textFormat, TextFormat.Getter<C> getter, C carrier, Q request) {
+    checkNotNull(textFormat, "textFormat");
+    checkNotNull(getter, "getter");
+    checkNotNull(carrier, "carrier");
+    checkNotNull(request, "request");
+    SpanBuilder spanBuilder = null;
+    String spanName = customizer.getSpanName(request, extractor);
+    String parseError = null;
+    // de-serialize the context
+    try {
+      SpanContext spanContext = textFormat.extract(carrier, getter);
+      spanBuilder = tracer.spanBuilderWithRemoteParent(spanName, spanContext);
+    } catch (Exception e) {
```

Comment:
Maybe it would make sense to be more explicit here what exceptions you catch.

`textFormat.extract` throws a `SpanContextParseException` when the input is invalid

`tracer.spanBuilderWithRemoteParent` throws a `NullPointerException` when the spanName is null.

I think those two error cases need explicit handling, especially when the second case happens the whole thing will blew up since `tracer.spanBuilder` will again throw a `NullPointerException`.

Reply:
Thanks. Refined the logic.
Currently we can't distinguish between parse error and missing context so I just ignore the exception. I added a TODO here.
As for the `NullPointerException`, we currently use a non-public customizer and it always return non-null span name. I added a TODO here as well.